pep-0639.rst

```diff
@@ -1,543 +1,2208 @@
+``License-File`` field.
+
+For backwards compatibility, to preserve consistent behavior with current tools
+and ensure that users do not unknowingly create packages that are not legally
+distributable, tools MUST assume the above default value for the
+``license-files`` key and also include, in addition to the license file
+specified under this ``file`` subkey, any license files that match the
+specified list of patterns.
+
+The ``license`` key may be removed from a new version of the specification
+in a future PEP.
+
+
+License files in project formats
+--------------------------------
+
+A few minor additions will be made to the relevant existing specifications
+to document, standardize and clarify what is already currently supported,
+allowed and implemented behavior, as well as explicitly mention the directory
+location the license file tree is rooted in for each format, per the
+`specification above <Add License-File field_>`_.
+
+**Project source trees**
+  As `described above <Project source metadata_>`_, the
+  `Declaring Project Metadata specification <#pep621spec_>`_
+  will be updated to reflect that license file paths MUST be relative to the
+  project root directory; i.e. the directory containing the ``pyproject.toml``
+  (or equivalently, other legacy project configuration,
+  e.g. ``setup.py``, ``setup.cfg``, etc).
+
+**Source distributions** *(sdists)*
+  The `sdist specification <#sdistspec_>`_ will be updated to reflect that for
+  ``Metadata-Version`` is ``2.3`` or greater, the sdist MUST contain any
+  license files specified by ``License-File`` in the ``PKG-INFO`` at their
+  respective paths relative to the top-level directory of the sdist
+  (containing the ``pyproject.toml`` and the ``PKG-INFO`` core metadata).
+
+**Built distributions** *(wheels)*
+  The `wheel specification <#wheelspec_>`_ will be updated to reflect that if
+  the ``Metadata-Version`` is ``2.3`` or greater and one or more
+  ``License-File`` fields is specified, the ``.dist-info`` directory MUST
+  contain a ``license_files`` subdirectory which MUST contain the files listed
+  in the ``License-File`` fields in the ``METADATA`` file at their respective
+  paths relative to the ``license_files`` directory.
+
+**Installed projects**
+  The `Recording Installed Projects specification <#installedspec_>`_ will be
+  updated to reflect that if the ``Metadata-Version`` is ``2.3`` or greater
+  and one or more ``License-File`` fields is specified, the ``.dist-info``
+  directory MUST contain a ``license_files`` subdirectory which MUST contain
+  the files listed in the ``License-File`` fields in the ``METADATA`` file
+  at their respective paths relative to the ``license_files`` directory,
+  and that any files in this directory MUST be copied from wheels
+  by install tools.
+
+
+Converting legacy metadata
+--------------------------
+
+If the contents of the ``License`` field are a valid license expression
```

Comment:
Great point; this section was originally written by the original author when the PEP was solely about updating the core metadata spec and wasn't concerned with PEP 621 keys. I do discuss it in that section and link this one from there, but I agree it would make sense to include this more explicitly. And actually, instead of imply that tools should manipulate the already-rendered distribution metadata, it makes sense to reframe this section to discuss this from the perspective of the project source metadata instead, and so I updated it accordingly.

This also reminds me, though, of a big reason I don't think I explicitly mentioned to @ofek and @uranusjr before as to why `license.text` and `license-expression` had to go under different top-level keys—otherwise, due to the rules on `dynamic`, it would be impossible to do any conversion if the field was defined at all, defeating much of the purpose of it.